# DN - priprava in čiščenje podatkov

Skozi domačo nalogo bomo uporabljali dataset oglasov rabljenih avtomobilov pridobljenih iz nemškega *eBay Kleinanzeigen*.

Podatki so razvrščeni v sledeče stolpce:
* *dateCrawled* - When this ad was first crawled. All field-values are taken from this date.
* *name* - Name of the car.
* *seller* - Whether the seller is private or a dealer.
* *offerType* - The type of listing
* *price* - The price on the ad to sell the car.
* *abtest* - Whether the listing is included in an A/B test.
* *vehicleType* - The vehicle Type.
* *yearOfRegistration* - The year in which the car was first registered.
* *gearbox* - The transmission type.
* *powerPS* - The power of the car in PS.
* *model* - The car model name.
* *kilometer* - How many kilometers the car has driven.
* *monthOfRegistration* - The month in which the car was first registered.
* *fuelType* - What type of fuel the car uses.
* *brand* - The brand of the car.
* *notRepairedDamage* - If the car has a damage which is not yet repaired.
* *dateCreated* - The date on which the eBay listing was created.
* *nrOfPictures* - The number of pictures in the ad.
* *postalCode* - The postal code for the location of the vehicle.
* *lastSeenOnline* - When the crawler saw this ad last online.

Tekom domače naloge bomo podatke očistili ter opravili neko osnovno analizo.

In [90]:
import pandas as pd
import numpy as np

autos = pd.read_csv('data/autos.csv', encoding='Latin-1')

<div class="alert alert-block alert-info">
<b>Vaja: </b> Uporabite .info() metodo za izpis osnovnih informacij o našem dataframe-u.
</div>

In [91]:
# Rešitev
autos.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [92]:
print(f"dimenzije dataframe-a: št. vrstic je {autos.shape[0]}, št. stolpcev je {autos.shape[1]}")

dimenzije dataframe-a: št. vrstic je 50000, št. stolpcev je 20


<div class="alert alert-block alert-info">
<b>Vaja: </b>Izpišite prvih 5 vrstic dataframe-a.
</div>

In [93]:
# Rešitev
autos.head(5)


dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Iz zgornjih vrstic lahko vidimo, da:
* Dataset ima 20 stolpcev, večina katerih je tipa string
* Nekaj stolpcev ima null vrednosti

Za začetek bomo poiskali stolpce v katerih je večina vrednosti istih, ker taki stolpci za analizo ponavadi niso koristni.

<div class="alert alert-block alert-info">
<b>Vaja: </b>Uporabite metodo .describe() in izpišite osnovne statistike vseh 20 stolpcev
</div>

In [94]:
# Rešitev
autos.describe(include='all')


dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

Opazimo naslednje stvari:
* stolpca *seller* in *offerType* imata skoraj vse vrednosti enake. Ta dva stolpca bomo odstranili
* vse vrednosti stolpca *nrOfPictures* so null. Stolpec bomo odstranilI
* *price* stolpec ima vrednosti v obliki string. To bomo spremenili v številčno vrednost
* *odometer* stolpec ima vrednosti v obliki string. To bomo spremenili v številčno vrednost

<div class="alert alert-block alert-info">
<b>Vaja: </b>Odstranite stolpce seller, offerType in nrOfPictuers.
</div>

In [95]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [96]:
# Rešitev
autos.drop(columns=['seller', 'offerType', 'nrOfPictures'], inplace=True)
autos.shape


(50000, 17)

<div class="alert alert-block alert-info">
<b>Vaja: </b>Vrednosti v stolpcu price so tipa string. Odstranite nepotrebne znake ($) ter vrednosti pretvorite v tip integer.
</div>

In [97]:
# Rešitev
autos["price"] = autos["price"].str.replace('$', '').str.replace(",", "").astype(int)

C:\Users\matic.zirovec\AppData\Local\Temp\ipykernel_15768\1135492678.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace('$', '').str.replace(",", "").astype(int)


In [98]:
autos.describe(include='all')

dateCrawled         name         price abtest vehicleType  \
count                 50000        50000  5.000000e+04  50000       44905   
unique                48213        38754           NaN      2           8   
top     2016-04-02 11:37:04  Ford_Fiesta           NaN   test   limousine   
freq                      3           78           NaN  25756       12859   
mean                    NaN          NaN  9.840044e+03    NaN         NaN   
std                     NaN          NaN  4.811044e+05    NaN         NaN   
min                     NaN          NaN  0.000000e+00    NaN         NaN   
25%                     NaN          NaN  1.100000e+03    NaN         NaN   
50%                     NaN          NaN  2.950000e+03    NaN         NaN   
75%                     NaN          NaN  7.200000e+03    NaN         NaN   
max                     NaN          NaN  1.000000e+08    NaN         NaN   

        yearOfRegistration  gearbox       powerPS  model   odometer  \
count         50000.000000    47320  50000.000000  47242      50000   
unique                 NaN        2           NaN    245         13   
top                    NaN  manuell           NaN   golf  150,000km   
freq                   NaN    36993           NaN   4024      32424   
mean           2005.073280      NaN    116.355920    NaN        NaN   
std             105.712813      NaN    209.216627    NaN        NaN   
min            1000.000000      NaN      0.000000    NaN        NaN   
25%            1999.000000      NaN     70.000000    NaN        NaN   
50%            2003.000000      NaN    105.000000    NaN        NaN   
75%            2008.000000      NaN    150.000000    NaN        NaN   
max            9999.000000      NaN  17700.000000    NaN        NaN   

        monthOfRegistration fuelType       brand notRepairedDamage  \
count          50000.000000    45518       50000             40171   
unique                  NaN        7          40                 2   
top                     NaN   benzin  volkswagen              nein   
freq                    NaN    30107       10687             35232   
mean               5.723360      NaN         NaN               NaN   
std                3.711984      NaN         NaN               NaN   
min                0.000000      NaN         NaN               NaN   
25%                3.000000      NaN         NaN               NaN   
50%                6.000000      NaN         NaN               NaN   
75%                9.000000      NaN         NaN               NaN   
max               12.000000      NaN         NaN               NaN   

                dateCreated    postalCode             lastSeen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

<div class="alert alert-block alert-info">
<b>Vaja: </b>Vrednosti v stolpcu odometer prav tako pretvorite v integer vrednosti.
</div>

In [99]:
# Rešitev
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "").astype(int)

In [100]:
autos.describe()

price  yearOfRegistration       powerPS       odometer  \
count  5.000000e+04        50000.000000  50000.000000   50000.000000   
mean   9.840044e+03         2005.073280    116.355920  125732.700000   
std    4.811044e+05          105.712813    209.216627   40042.211706   
min    0.000000e+00         1000.000000      0.000000    5000.000000   
25%    1.100000e+03         1999.000000     70.000000  125000.000000   
50%    2.950000e+03         2003.000000    105.000000  150000.000000   
75%    7.200000e+03         2008.000000    150.000000  150000.000000   
max    1.000000e+08         9999.000000  17700.000000  150000.000000   

       monthOfRegistration    postalCode  
count         50000.000000  50000.000000  
mean              5.723360  50813.627300  
std               3.711984  25779.747957  
min               0.000000   1067.000000  
25%               3.000000  30451.000000  
50%               6.000000  49577.000000  
75%               9.000000  71540.000000  
max              12.000000  99998.000000

Sedaj si bomo pogledali stolpca *price* in *odometer* bolj podrobno.

<div class="alert alert-block alert-info">
<b>Vaja: </b>Izpišite maximalno in minimalno vrednost stolpca price ter kolikokrat se te dve vrednosti pojavita.
</div>

In [101]:
# Rešitev
max_min_values = autos.describe().loc[["max", "min"],["price", "odometer"]]

max_price = max_min_values.loc["max", "price"]
min_price = max_min_values.loc["min", "price"]

print(f"Max price={max_price}, st.pojavitev={ autos['price'].value_counts().loc[max_price]}")
print(f"Max price={min_price}, st.pojavitev={ autos['price'].value_counts().loc[min_price]}")

max_odometer = max_min_values.loc["max", "odometer"]
min_odometer = max_min_values.loc["min", "odometer"]

print(f"Max odometer={max_odometer}, st.pojavitev={ autos['odometer'].value_counts().loc[max_odometer]}")
print(f"Max odometer={min_odometer}, st.pojavitev={ autos['odometer'].value_counts().loc[min_odometer]}")




Max price=99999999.0, st.pojavitev=1
Max price=0.0, st.pojavitev=1421
Max odometer=150000.0, st.pojavitev=32424
Max odometer=5000.0, st.pojavitev=967


<div class="alert alert-block alert-info">
<b>Vaja: </b>Obdržite vrstice kjer je vrednost price stolpca med vključno 1 in 351000.
</div>

In [102]:
# Rešitev
# autos.shape
autos = autos.loc[(autos["price"]>=1) & (autos["price"]<=351000),:]

Poglejmo si sedaj stolpec *yearOfRegistration*.

<div class="alert alert-block alert-info">
<b>Vaja: </b>Poiščite max in min vrednost stolpca yearOfRegistration.
</div>

In [103]:
# Rešitev
autos.describe().loc[["max", "min"],["yearOfRegistration"]]

yearOfRegistration
max              9999.0
min              1000.0

Opazimo, da je minimalna vrednost 1000 (daleč predno so iznašli prvi avto), in maximalna vrednost 9999 (daleč v prihodnosti).

Ker avto ne more biti prvič registriran šele potem, ko je objavljen oglas, so vse letnice nad 2016 nepravilne.

Težje je določiti začetno veljavno leto. Realistično je to lahko nekje v prvih desetletjih 1900. 

<div class="alert alert-block alert-info">
<b>Vaja: </b>Obdržite vrstice, kjer se vrednost stolpca yearOfRegistration nahaja znotraj območja 1900 - 2016.
</div>

In [104]:
# Rešitev
autos = autos.loc[autos["yearOfRegistration"].isin(range(1900,2016+1)),:]


Sedaj si bomo pogledali stolpca *price* in *brand*.

<div class="alert alert-block alert-info">
<b>Vaja: </b>Izpišite katere vrednosti se nahajajo v stolpcu brand in kolikokrat se vsaka vrednost pojavi (v %).
</div>

In [105]:
# Rešitev
autos["brand"].value_counts(normalize=True)*100


volkswagen        21.126368
bmw               11.004477
opel              10.758124
mercedes_benz      9.646323
audi               8.656627
ford               6.989996
renault            4.714980
peugeot            2.984083
fiat               2.564212
seat               1.827296
skoda              1.640925
nissan             1.527388
mazda              1.518819
smart              1.415994
citroen            1.400998
toyota             1.270324
hyundai            1.002549
sonstige_autos     0.981127
volvo              0.914719
mini               0.876159
mitsubishi         0.822604
honda              0.784045
kia                0.706926
alfa_romeo         0.664082
porsche            0.612669
suzuki             0.593389
chevrolet          0.569825
chrysler           0.351321
dacia              0.263490
daihatsu           0.250637
jeep               0.227073
subaru             0.214220
land_rover         0.209936
saab               0.164949
jaguar             0.156381
daewoo             0

Opazimo, da nemški proizvajalci predstavljajo štiri od pvih 5 blagovnih znamk. Vidimo, da skupaj predstavljajo več kot 50% vseh oglasov. Volkswagen je najbolj priljubljena znamka, s približno dvakrat večjim številom, kot naslednja znamka.

V nalogi se bomo osredotočili na znamke, ki predstavljajo več kot 5% oglasov.

<div class="alert alert-block alert-info">
<b>Vaja: </b>Obdržite vse znamke, ki se pojavijo več kot 5%.
</div>

In [106]:
# Rešitev
# (40,0)
brands = autos["brand"].value_counts(normalize=True)
frequent_brands = brands[brands> 0.05].index
autos = autos.loc[autos["brand"].isin(frequent_brands),:]


<div class="alert alert-block alert-info">
<b>Vaja: </b>Ustvarite nov dictionary in vanj shranite povprečno ceno za vsako znamko.
    
    "znamka": povprečna_cena,
    "znamka": povprečna_cena,
    "znamka": povprečna_cena,
    
</div>

In [116]:
# Rešitev
avg_prices = autos.groupby("brand")["price"].mean().sort_values(ascending=False).to_dict()
print(avg_prices)


{'audi': 9336.687453600594, 'mercedes_benz': 8628.450366422385, 'bmw': 8332.820517811953, 'volkswagen': 5402.410261610221, 'ford': 3749.4695065890287, 'opel': 2975.2419354838707}


Vidimo, da je izrazit razkorak v cenah:
* Audi, BMW in Marcedes Benz so dražje znamke
* For in Opel sta cenejši
* Volkswagen je nekje vmes

<div class="alert alert-block alert-info">
<b>Vaja: </b>Iz zgornjega dictionary ustvarite nov pandas dataframe, kjer je ime znamke index vrstice in povprečna cena stolpec.
</div>

In [131]:
# Rešitev
s = pd.Series(avg_prices)
df = pd.concat([s], axis=1).rename(columns={0:"povprecna_cena"})
df

povprecna_cena
audi              9336.687454
mercedes_benz     8628.450366
bmw               8332.820518
volkswagen        5402.410262
ford              3749.469507
opel              2975.241935